rust_src/src/minibuf.rs

```diff
@@ -424,4 +427,107 @@ pub fn read_no_blanks_input(
     }
 }
 
+/// Read the name of a buffer and return as a string.
+/// Prompt with PROMPT.
+/// Optional second arg DEF is value to return if user enters an empty line.
+/// If DEF is a list of default values, return its first element.
+///
+/// Optional third arg REQUIRE-MATCH determines whether non-existing
+/// buffer names are allowed.  It has the same meaning as the
+/// REQUIRE-MATCH argument of `completing-read'.
+///
+/// The argument PROMPT should be a string ending with a colon and a space.
+/// If `read-buffer-completion-ignore-case' is non-nil, completion ignores
+/// case while reading the buffer name.
+/// If `read-buffer-function' is non-nil, this works by calling it as a
+/// function, instead of the usual behavior.
+///
+/// Optional arg PREDICATE if non-nil is a function limiting the buffers that can
+/// be considered.
+#[lisp_fn(min = "1")]
+pub fn read_buffer(
+    mut prompt: Option<LispStringRef>,
+    mut def: LispObject,
+    require_match: LispObject,
+    predicate: LispObject,
+) -> LispObject {
+    let count = c_specpdl_index();
+
+    if let Some(buffer) = def.as_buffer() {
+        def = buffer.name();
+    }
+
+    unsafe {
+        specbind(
+            Qcompletion_ignore_case,
+            globals.read_buffer_completion_ignore_case.into(),
+        )
+    }
+
+    let result = if unsafe { globals.Vread_buffer_function }.is_nil() {
+        if !def.is_nil() {
+            // A default was provided: PROMT must be changed, editing in the
+            // default value before the colon. To achieve this, PROMT is replaced
+            // with a substring that doesn't contain the terminal space and colon
+            // (if present).
+            if let Some(mut string) = prompt {
+                let data = string.as_slice();
+                let mut len = string.len_bytes() as usize;
+                if len >= 2 && &data[len - 2..len] == &[b':', b' '] {
+                    len -= 2;
+                } else if len >= 1 && (data[len - 1] == b':' || data[len - 1] == b' ') {
+                    len -= 1;
+                }
+                prompt = unsafe {
+                    make_specified_string(
+                        string.sdata_ptr(),
+                        -1,
+                        len as isize,
+                        string.is_multibyte(),
+                    )
+                    .into()
+                };
+            }
+            let format_str = CString::new("%s (default %s): ").unwrap();
```

Comment:
These 3 lines could be turned into a macro easily enough. `let format = new_unibyte_string!("%s (default %s): ");`